crates/net/eth-wire/src/types/message.rs

```diff
@@ -237,6 +237,143 @@ impl EthMessage {
+    /// * `approx` - The approximate size to allocate in memory for the encoding process. This is
+    ///   used to pre-allocate memory for efficiency.
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A vector containing the encoded data, truncated to fit within the size limit if necessary.
+    /// Users of this method should be aware that the truncated data may not represent the complete
+    /// original dataset.
+    fn encode_truncate(&self, approx: usize, limit: usize) -> Vec<u8>;
+
+    /// Encodes the data up to the specified size limit, returning the encoded data along with
+    /// a flag indicating whether the limit was exceeded.
+    ///
+    /// # Arguments
+    ///
+    /// * `approx` - The approximate size to allocate in memory for the encoding process. This is
+    ///   used to pre-allocate memory for efficiency.
+    /// * `limit` - The maximum allowed size of the encoded data.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing:
+    /// * The encoded data as `Vec<u8>`.
+    /// * The length of the data at the point just before reaching the limit, or 0 if the limit was
+    ///   not exceeded.
+    fn encode_until_limit(&self, approx: usize, limit: usize) -> (Vec<u8>, usize);
+}
+
+impl<T: Encodable> EncodableExt for Vec<T> {
+    fn encode_max(
+        &self,
+        approx: usize,
+        limit: usize,
+    ) -> alloy_rlp::Result<Vec<u8>, alloy_rlp::Error> {
+        let (mut buf, current_len) = self.encode_until_limit(approx, limit);
+
+        if current_len != 0 {
+            // Don't shrink if limit_exceeded since we return Error anyway
+            return Err(alloy_rlp::Error::Custom("Size limit exceeded"));
+        }
+
+        buf.shrink_to_fit();
+        Ok(buf)
+    }
+
+    fn encode_until_limit(&self, approx: usize, limit: usize) -> (Vec<u8>, usize) {
+        let mut buffer = Vec::with_capacity(approx);
+        let mut header = Header { list: true, payload_length: 0 };
+        let mut actual_payload_length = 0;
+        let mut current_len = 0;
+        for item in self {
+            item.encode(&mut buffer);
+            actual_payload_length += item.length();
+            // retroactively check if current buffer , and the list header would be > limit
+            if check_limit(&buffer, actual_payload_length) > limit {
+                let mut head: Vec<u8> = Vec::new();
+                header.payload_length = actual_payload_length - item.length();
+                header.encode(&mut head);
+                current_len += head.len();
+                head.append(&mut buffer);
+                return (head.clone(), current_len);
+            }
+
+            current_len = buffer.len();
+        }
+        let mut head: Vec<u8> = Vec::new();
+        header.payload_length = actual_payload_length;
+        header.encode(&mut head);
+        head.append(&mut buffer);
+        // Return current_len as 0
+        (head, 0)
```

Comment:
this looks very complex,

I don't think this is necessary if we first check how much items we can fit into the buffer

ptal at the solution @Rjected suggested

Reply:
need a clarification 
by limit we mean the list header + the encoding right ? 
If yes, then , header length changes depending on what the payload_length is , ex: if header.payload_length = 0 , encoded list header's len will be 1 , and 3 in other cases iirc . That is why I am calculating the list header length above.
Ex: if expected payload length is 508, limit is 511 , the `if ` condition will not pass (since 508 is not > 511), but If we include the list header len(3 length as payload_length = 508), then it will pass.

And If you mean `limit` excludes the list header, then we can indeed calculate how many items will fit beforehand correctly.

Reply:
the algorithm is basically:
you have a limit
you have a list of encodable items
you have a "current" _list_ length and item (you could track an index)
for each item:
* new_list_length = current_list_length + item.length()
* if Header { list: true, new_list_length }.length() + new_list_length > limit, return
* otherwise proceed

Once you know which element you need to stop at, now you should encode, you encode the list of elements up to that item